server/channels/app/post.go

```diff
@@ -766,22 +766,27 @@ func (a *App) publishWebsocketEventForPost(c request.CTX, post *model.Post, mess
 
 	defer a.Publish(message)
 
-	var previewedPostID string
-	if val, ok := post.GetProp(model.PostPropsPreviewedPost).(string); ok {
-		previewedPostID = val
-	} else {
+	permalinkPreviewedPost := post.GetPreviewPost()
+	if permalinkPreviewedPost == nil {
 		return nil
 	}
 
-	if !model.IsValidId(previewedPostID) {
-		mlog.Warn("invalid post prop value", mlog.String("prop_key", model.PostPropsPreviewedPost), mlog.String("prop_value", previewedPostID))
+	if !model.IsValidId(permalinkPreviewedPost.PostID) {
+		mlog.Warn("invalid preview post ID", mlog.String("prop_value", permalinkPreviewedPost.PostID))
 		return nil
 	}
 
-	previewedPost, appErr := a.GetSinglePost(previewedPostID, false)
+	// To remain secure by default, we wipe out the metadata unconditionally.
+	post.Metadata.Embeds[0].Data = nil
+	postWithoutPermalinkPreviewJSON, err := post.ToJSON()
+	if err != nil {
+		return model.NewAppError("publishWebsocketEventForPost", "app.post.marshal.app_error", nil, "", http.StatusInternalServerError).Wrap(jsonErr)
+	}
+
+	previewedPost, appErr := a.GetSinglePost(permalinkPreviewedPost.PostID, false)
```

Comment:
More thoughts -- unrelated to this PR, but while we're here! -- why do we have to `a.GetSinglePost` and `a.GetChannel` at all? The `PreviewPost` contains all the information we need to do the permission checks. So it just comes down to whether or not we trust the caller not to provide bogus data?

Reply:
We can do away with `GetSinglePost`, but we need `GetChannel` to get the channel.TeamId for the `a.HasPermissionToReadChannel` check.